src/hyperlight_host/src/hypervisor/gdb/hyp_debug.rs

```diff
@@ -0,0 +1,758 @@
+/// Check page 7-28 Vol. 3A of Intel 64 and IA-32
+/// Architectures Software Developer's Manual
+const SW_BP_OP: u8 = 0xCC;
+/// Software Breakpoint written to memory
+pub const SW_BP: [u8; SW_BP_SIZE] = [SW_BP_OP];
+/// Exception id for SW breakpoint
+const SW_BP_ID: u32 = 3;
+/// Max number of supported hw breakpoints
+const MAX_NO_OF_HW_BP: usize = 4;
+
+/// Check page 19-4 Vol. 3B of Intel 64 and IA-32
+/// Architectures Software Developer's Manual
+/// Bit position of BS flag in DR6 debug register
+const DR6_BS_FLAG_POS: usize = 14;
+/// Bit mask of BS flag in DR6 debug register
+const DR6_BS_FLAG_MASK: u64 = 1 << DR6_BS_FLAG_POS;
+/// Bit position of HW breakpoints status in DR6 debug register
+const DR6_HW_BP_FLAGS_POS: usize = 0;
+/// Bit mask of HW breakpoints status in DR6 debug register
+const DR6_HW_BP_FLAGS_MASK: u64 = 0x0F << DR6_HW_BP_FLAGS_POS;
+
+/// This trait is used to define a common way of interacting with a vCPU to
+/// allow debugging functionality
+pub trait GuestVcpuDebug {
+    /// Type that wraps the vCPU functionality
+    type Vcpu;
+
+    /// Returns true whether the provided address is a hardware breakpoint
+    fn is_hw_breakpoint(&self, addr: &u64) -> bool;
+    /// Stores the address of the hw breakpoint
+    fn save_hw_breakpoint(&mut self, addr: &u64) -> bool;
+    /// Deletes the address of the hw breakpoint from storage
+    fn delete_hw_breakpoint(&mut self, addr: &u64);
+
+    /// Read registers
+    fn read_regs(&self, vcpu_fd: &Self::Vcpu, regs: &mut X86_64Regs) -> Result<()>;
+    /// Enables or disables stepping and sets the vCPU debug configuration
+    fn set_single_step(&mut self, vcpu_fd: &Self::Vcpu, enable: bool) -> Result<()>;
+    /// Translates the guest address to physical address
+    fn translate_gva(&self, vcpu_fd: &Self::Vcpu, gva: u64) -> Result<u64>;
+    /// Write registers
+    fn write_regs(&self, vcpu_fd: &Self::Vcpu, regs: &X86_64Regs) -> Result<()>;
+
+    /// Adds hardware breakpoint
+    fn add_hw_breakpoint(&mut self, vcpu_fd: &Self::Vcpu, addr: u64) -> Result<bool> {
+        let addr = self.translate_gva(vcpu_fd, addr)?;
+
+        if self.is_hw_breakpoint(&addr) {
+            Ok(true)
+        } else {
+            let res = self.save_hw_breakpoint(&addr);
+            if res {
+                self.set_single_step(vcpu_fd, false)?;
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        }
+    }
+    /// Removes hardware breakpoint
+    fn remove_hw_breakpoint(&mut self, vcpu_fd: &Self::Vcpu, addr: u64) -> Result<bool> {
+        let addr = self.translate_gva(vcpu_fd, addr)?;
+
+        if self.is_hw_breakpoint(&addr) {
+            self.delete_hw_breakpoint(&addr);
+            self.set_single_step(vcpu_fd, false)?;
+
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+}
+
+/// Trait used for defining memory debugging functionality
+pub trait GuestMemoryDebug: GuestVcpuDebug {
```

Comment:
Why split this into a separate trait? Also–is naming appropriate? If so, might be worth further elaborating in the comment why memory debugging = software breakpoints and vCPU debugging = hardware breakpoints.

Reply:
This is more of a logical split, rather than functional, meant to allow only vCPU manipulation without memory access.
Indeed, I don't think we'll ever have a use case where we don't allow memory manipulation of the guest.
So, I'll merge them together and rename the trait

Reply:
Refactored and merged them together.